### PR TITLE
support bitsandbytes 8-bit and FP4 quantized models

### LINKED ISSUE #6756: [Bug]: Unable to run meta-llama/Llama-Guard-3-8B-INT8
### Your current environment

Latest Docker image, RTX 4090

### 🐛 Describe the bug

```
docker run --gpus all vllm/vllm-openai:latest --model meta-llama/Llama-Guard-3-8B-INT8
...
[rank0]:     raise ValueError(f"Cannot find any of {keys} in the model's "
[rank0]: ValueError: Cannot find any of ['adapter_name_or_path'] in the model's quantization config.
```

### PR DESCRIPTION
**Pull request by chenqianfzh**
This PR does the following:
1. support quantized bitsandbytes 8-bit models, such as meta-llama/Llama-Guard-3-8B-INT8
2. support quantized bitsandbytes 4-bit FP4 models, such as PrunaAI/Einstein-v6.1-Llama3-8B-bnb-4bit-smashed
3. Add comments about enforcing eager-mode in bnb quantization, as I identified it is a bug in the underlying dependency package of bitsandbytes. 

FIX https://github.com/vllm-project/vllm/issues/6756

### CODE REVIEW

File: tests/quantization/test_bitsandbytes.py [hunk 1]
```diff
@@ -5,82 +5,108 @@
...
+        actual_output = outputs[index][1][0].split('\n', 1)[0]
+        expected_output = expected_outputs[index].split('\n', 1)[0]
+
+        assert len(actual_output) >= len(expected_output), (
+            f'Actual {actual_output} should be larger than or equal to '
+            f'expected {expected_output}')
+        actual_output = actual_output[:len(expected_output)]
+
+        assert actual_output == expected_output, (
+            f'Expected: {expected_output}, but got: {actual_output}')
```
Comment (mgoin) [inline tests/quantization/test_bitsandbytes.py:112]: Could you change this to do a direct top-logprob greedy comparison between HF and vLLM? There are several references in the testing but here is one https://github.com/vllm-project/vllm/blob/9ba85bc1527928f30ec56961520d7e07ee385167/tests/models/test_mistral.py#L29-L35

**Review by mgoin — COMMENTED** [review]
This is great work and refactoring, appreciate it! I need to do another pass through as it's a bit dense, so if you could document more of the config arguments and _apply_8bit_weight that would be helpful

**Comment by chenqianfzh** [comment]
> This is great work and refactoring, appreciate it! I need to do another pass through as it's a bit dense, so if you could document more of the config arguments and _apply_8bit_weight that would be helpful

Thanks. Will do.

File: tests/quantization/test_bitsandbytes.py (hunk 1 above)
Comment (chenqianfzh) [reply tests/quantization/test_bitsandbytes.py]: > Could you change this to do a direct top-logprob greedy comparison between HF and vLLM? There are several references in the testing but here is one

Updated the code per comments, could you take a look?

File: tests/quantization/test_bitsandbytes.py [hunk 2]
```diff
@@ -2,85 +2,168 @@
...
+    # Clean up the GPU memory for the next test
+    gc.collect()
+    torch.cuda.empty_cache()
+
+    #Run with vLLM runner
     with vllm_runner(model_name,
                      quantization='bitsandbytes',
                      load_format='bitsandbytes',
-                     enforce_eager=True) as llm:
+                     enforce_eager=True,
```
Comment (jeejeelee) [inline tests/quantization/test_bitsandbytes.py:103]: https://github.com/bitsandbytes-foundation/bitsandbytes/pull/1330 has been merged. Regarding these tests, can we now use cudagraph?

Comment (chenqianfzh) [reply tests/quantization/test_bitsandbytes.py:103]: Hi, @jeejeelee, thanks to your fix in bnb. 

But the latest version of bnb package was released three weeks ago, which does not include your fix yet. 

I will update the code after the next bnb release is out.

Comment (mgoin) [reply tests/quantization/test_bitsandbytes.py:103]: Agreed, I think it is worth doing the package upgrade in another PR

**Review by mgoin — APPROVED** [review]
Thanks for the improvements, this looks good to me

File: tests/quantization/test_bitsandbytes.py (hunk 2 above)
Comment (jeejeelee) [reply tests/quantization/test_bitsandbytes.py:103]: It seems that this PR introduces a new BNB kernel . I'm not sure if the previous modifications to BNB can support this kernel. What I mean is, perhaps we should first verify this (build BNB from source). If this kernel still not supported, we may need to continue refining the relevant BNB code.

If you're not available, I can verify it next week.

Comment (chenqianfzh) [reply tests/quantization/test_bitsandbytes.py:103]: @jeejeelee I tried the new BNB kernel with your fix. Found the current BNB main branch is suffering some regression.

I got the following error in graph mode as well as eager mode:

[rank0]:   File "/home/ubuntu/qian/git/vllm/vllm/model_executor/model_loader/loader.py", line 893, in _unquantized_generator
[rank0]:     processed_weight, quant_state = quantize_4bit(
[rank0]:                                     ^^^^^^^^^^^^^^
[rank0]:   File "/home/ubuntu/qian/git/bitsandbytes/bitsandbytes/functional.py", line 1218, in quantize_4bit
[rank0]:     lib.cquantize_blockwise_fp16_nf4(
[rank0]:     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]: AttributeError: 'NoneType' object has no attribute 'cquantize_blockwise_fp16_nf4'

I commented on the bnb bug thread of https://github.com/bitsandbytes-foundation/bitsandbytes/issues/1210 and look forward for some insights from the community.

**Comment by chenqianfzh** [comment]
@mgoin The test errors seem unrelated to my change. What shall I do?

**Comment by mgoin** [comment]
@chenqianfzh could you please merge with latest main? recent PRs don't seem to be failing, so I wouldn't expect test errors

**Comment by mgoin** [comment]
I'm not sure what is the issue, I was also manually retrying tests in buildkite.. I will run the tests locally

**Comment by chenqianfzh** [comment]
> I'm not sure what is the issue, I was also manually retrying tests in buildkite.. I will run the tests locally

My local tests always pass. 

However, I guess it is related to GPU memories not release correctly. I am trying my fix with a fake PR now.

**Comment by chenqianfzh** [comment]
@mgoin I found all the checks have passed, could you help merge it? Thx

**Comment by jvlinsta** [comment]
Does this also mean we can use bitsandbytes with tensor-parallel-size > 1?

**Comment by jvlinsta** [comment]
Hi @chenqianfzh thanks for that! Where does that PR live, so I can keep following up on it? ^^

**Comment by jvlinsta** [comment]
Is it here? https://github.com/bd-iaas-us/vllm/commit/e8d54539e1f4d5373c9f8d20969decd628f86816

**Comment by chenqianfzh** [comment]
> Is it here? [bd-iaas-us@e8d5453](https://github.com/bd-iaas-us/vllm/commit/e8d54539e1f4d5373c9f8d20969decd628f86816)

yep. https://github.com/vllm-project/vllm/pull/8434 is for the bnb TP.

**Pull request merged**